lighthouse-core/gather/driver/network-monitor.js

```diff
@@ -0,0 +1,241 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const log = require('lighthouse-logger');
+const {EventEmitter} = require('events');
+const NetworkRecorder = require('../../lib/network-recorder.js');
+const {NON_NETWORK_PROTOCOLS} = require('../../lib/url-shim.js');
+
+/** @typedef {import('../../lib/network-recorder.js').NetworkRecorderEvent} NetworkRecorderEvent */
+/** @typedef {import('../../lib/network-request.js')} NetworkRequest */
+/** @typedef {'network-2-idle'|'network-critical-idle'|'networkidle'|'networkbusy'|'network-critical-busy'|'network-2-busy'} NetworkMonitorEvent_ */
+/** @typedef {NetworkRecorderEvent|NetworkMonitorEvent_} NetworkMonitorEvent */
+
+const IGNORED_NETWORK_SCHEMES = [...NON_NETWORK_PROTOCOLS, 'ws'];
+
+/**
+ * @fileoverview This class wires up the procotol to a network recorder and provides overall
+ * status inspection state.
+ */
+
+class NetworkMonitor extends EventEmitter {
+  /** @type {NetworkRecorder|undefined} */
+  _networkRecorder = undefined;
+  /** @type {Array<LH.Crdp.Page.Frame>} */
+  _frameNavigations = [];
+
+  // Redefine the event emitter types with a narrower type signature.
+  /** @param {NetworkMonitorEvent} event @param {*} listener  */
+  on = (event, listener) => super.on(event, listener);
+  /** @param {NetworkMonitorEvent} event @param {*} listener  */
+  once = (event, listener) => super.once(event, listener);
+  /** @param {NetworkMonitorEvent} event @param {*} listener  */
+  off = (event, listener) => super.off(event, listener);
+
+  /** @param {LH.Gatherer.FRProtocolSession} session */
+  constructor(session) {
+    super();
+    this._session = session;
+
+    this._session.on('Page.frameNavigated', event => {
+      if (!this._networkRecorder) return;
+      this._frameNavigations.push(event.frame);
+    });
+
+    this._session.onAny(event => {
+      if (!this._networkRecorder) return;
+      this._networkRecorder.dispatch(event);
+    });
+  }
+
+  /**
+   * @return {Promise<void>}
+   */
+  async enable() {
+    if (this._networkRecorder) return;
+
+    this._frameNavigations = [];
+    this._networkRecorder = new NetworkRecorder();
+
+    /**
```

Comment:
this is sorta new, we're reemitting the request start/end and the `this._emitNetworkStatus` was moved from recorder to here